tools/mac-distribution/packaging/appbundle.go

```diff
@@ -0,0 +1,112 @@
+package packaging
+
+import (
+	"fmt"
+	"log/slog"
+	"os"
+	"path/filepath"
+
+	"github.com/gravitational/shared-workflows/tools/mac-distribution/internal/fileutil"
+	"github.com/gravitational/shared-workflows/tools/mac-distribution/notarize"
+	"github.com/gravitational/trace"
+)
+
+// AppBundlePackager is a packager for creating an app bundle (.app) for distribution.
+type AppBundlePackager struct {
+	Info *AppBundleInfo
+
+	log        *slog.Logger
+	notaryTool *notarize.Tool
+}
+
+// AppBundleInfo represents an app bundle to be packaged for distribution.
+type AppBundleInfo struct {
+	// Skeleton directory to use as the base for the app bundle
+	Skeleton string
+	// Entitlements file to use for the app bundle
+	Entitlements string
+	// AppBinary is the binary to use as the main executable for the app bundle
+	AppBinary string
+}
+
+// AppBundlePackagerOpts contains options for creating an AppBundlePackager.
+type AppBundlePackagerOpts struct {
+	// NotarizationEnabled determines whether to notarize the app bundle
+	NotaryTool *notarize.Tool
+	// Logger is the logger to use for the packager
+	Logger *slog.Logger
+}
+
+// NewAppBundlePackager creates a new AppBundlePackager.
+func NewAppBundlePackager(info *AppBundleInfo, opts *AppBundlePackagerOpts) *AppBundlePackager {
+	log := opts.Logger
+	if log == nil {
+		log = slog.Default()
+	}
+
+	return &AppBundlePackager{
+		Info:       info,
+		log:        log,
+		notaryTool: opts.NotaryTool,
+	}
+}
+
+// Package creates an app bundle from the provided info.
+func (a *AppBundlePackager) Package() error {
+	if err := a.Info.validate(); err != nil {
```

Comment:
Why are we validating this inside "Package()" rather than inside the NewAppBundlePackager "constructor"?